### PR TITLE
Turn missing storage locations into an error.

### PR DESCRIPTION
**Pull request by ekpyron**
~~Still needs updates for external tests.~~

### CODE REVIEW

**Comment by ekpyron** [comment]
Created a PR for solc-js changes: https://github.com/ethereum/solc-js/pull/236

File: docs/frequently-asked-questions.rst [hunk 1]
```diff
@@ -273,9 +273,11 @@ of variable it concerns:
 
 * state variables are always in storage
 * function arguments are in memory by default
```
Comment (chriseth) [inline docs/frequently-asked-questions.rst:275]: We remove the default here, too, don't we?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 2]
```diff
@@ -208,7 +208,7 @@ BOOST_AUTO_TEST_CASE(external_structs)
 			struct X { bytes32 x; Test t; Simple[] s; }
 			function f(ActionChoices, uint, Simple) external {}
 			function g(Test, Nested) external {}
-			function h(function(Nested) external returns (uint)[]) external {}
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:211]: Shouldn't this be `memory`?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 3]
```diff
@@ -236,7 +236,7 @@ BOOST_AUTO_TEST_CASE(external_structs_in_libraries)
...
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:239]: Same here.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:211]: I'm actually not sure.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/276_invalid_types_in_inline_array.sol [hunk 4]
```diff
@@ -1,8 +1,7 @@
 contract C {
     function f() public {
-        uint[3] x = [45, 'foo', true];
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/276_invalid_types_in_inline_array.sol:3]: I think this was rather meant to be `memory`.

**Review by chriseth — DISMISSED** [review]
Fine apart from the test changes.

File: docs/frequently-asked-questions.rst (hunk 1 above)
Comment (ekpyron) [reply docs/frequently-asked-questions.rst:275]: This PR doesn't - and there wasn't a warning about a location missing for those before. I was wondering, whether or not we should still remove the default here as well. I'll create an issue for it, if there isn't already one.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 2 above)
Comment (ekpyron) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:211]: I'm actually wondering right now, why this needs to be changed for this PR at all... Is a nested function argument somehow considered a local variable declaration? That would seem rather weird to me... I should probably look into that and at least understand why the code change above even affects this.

Comment (ekpyron) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:211]: The scope of the variable declaration ``Nested`` seems to be the callable declaration of function ``h``. And ``VariableDeclaration::isLocalVariable()`` considers everything a local variable, whose scope is a callable declaration, a block or a for statement.
Do you think this makes sense? To me it seems that this case should be excluded there and that ``Nested`` here is **not** a local variable. If you agree I'll create an issue for it.

Comment (ekpyron) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:211]: For this PR I now changed it to ``memory`` in any case, although (as it is considered a local variable, but **not** a callable parameter in ``VariableDeclaration::isCallableParameter``) it defaulted to ``storage`` before.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/276_invalid_types_in_inline_array.sol (hunk 4 above)
Comment (ekpyron) [reply test/libsolidity/syntaxTests/nameAndTypeResolution/276_invalid_types_in_inline_array.sol:3]: Yes, done.

**Review by chriseth — APPROVED** [review]

**Pull request merged**